pysyncobj/syncobj.py

```diff
@@ -172,6 +170,15 @@ def __init__(self, selfNode, otherNodes, conf=None, consumers=None, nodeClass =
         self.__destroying = False
         self.__recvTransmission = ''
 
+        if not self.__conf.journalFile or not self.__conf.flushJournal:
+            # Journal is either not written to a file or not flushed.
+            # We can't be sure whether this node voted in the current term yet, so don't vote until maxTimeout has elapsed.
+            # Factor 1.1 just to be safe.
+		# Note that an entire non-flushing cluster should always use the same maximum timeout to ensure that this method works!
+            self.__voteBlockTime = time.time() + 1.1 * self.__conf.raftMaxTimeout
```

Comment:
I can't see any example of using cluster where part of nodes has enabled flushing and part of them not. So there is no need to use this 1.1 constant. Also i'm not sure this gives any guarantee, looks like a hack.

Reply:
I don't see how having such a cluster relates to this change.

You're right that this factor 1.1 is a hack. The entire `voteBlockTime` thing is a giant hack really. The Raft algorithm requires that the term and vote (and the log) is stored to stable storage before responding to vote requests (or any other message), and using an in-memory journal or disabling flushing violates this requirement.

The Raft algorithm never relies on timing anywhere to guarantee consistency. Clock skew therefore won't affect reliability as long as the algorithm is implemented correctly. But in this particular case, clock skew *will* affect reliability (because the algorithm's assumptions are violated as mentioned above). For example, if the restarting node's clock is running faster than the rest of the cluster's clocks, it could still vote twice within a term. The factor 1.1 is a hacky fix for this issue in all but the most severe cases (if the clock is running more than 10 % too fast, something's *seriously* wrong).

In my opinion, using an in-memory journal or not flushing the journal file should not be an option in the first place since it violates a key assumption in Raft. I realise that this has a severe performance impact, but well, that's the cost for reliability. I prefer a slightly slower, reliable solution over a more performant one with nasty timing bugs that are nearly impossible to reproduce or debug.

Reply:
Ok, but may be it would be a good idea at least to move it to a config (instead of hardcoding in-place)?

Reply:
Sounds reasonable. I'll do that.